### PR TITLE
Bug 471434 - [contentassist][generator] the ContentAssist fragment should consider redefined grammar rules

### PR DESCRIPTION
**Pull request by LorenzoBettini**
I thought that it would have been nice to also have a test language, so I added one to the existing xbase.testlanguages project. I'm sorry this required to tweak a few things in that project and to regenerate also the existing language so this pull request contains many changes due to the regenerated files. Thus the first two commits are not important.

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=471434

**Comment by szarnekow** [comment]
Jenkins test

File: tests/org.eclipse.xtext.xbase.testlanguages.ui/plugin.xml [hunk 1]
```diff
@@ -9,7 +9,7 @@
             class="org.eclipse.xtext.xbase.testlanguages.ui.XImportSectionTestLangExecutableExtensionFactory:org.eclipse.xtext.ui.editor.XtextEditor"
             contributorClass="org.eclipse.ui.editors.text.TextEditorActionContributor"
             default="true"
-            extensions="__ximportimporttestlang"
+            extensions="ximportsectiontestlang"
```
Comment (szarnekow) [inline tests/org.eclipse.xtext.xbase.testlanguages.ui/plugin.xml]: This change looks bogus to me

File: tests/org.eclipse.xtext.xbase.ui.tests/src/org/eclipse/xtext/xbase/ui/tests/contentassist/Bug471434Test.xtend [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+package org.eclipse.xtext.xbase.ui.tests.contentassist
+
+import org.eclipse.xtext.xbase.junit.ui.AbstractContentAssistTest
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.eclipse.xtext.junit4.XtextRunner
+import org.eclipse.xtext.junit4.InjectWith
+import org.eclipse.xtext.xbase.testlanguages.ContentAssistFragmentTestLangUiInjectorProvider
+
+@RunWith(typeof(XtextRunner))
```
Comment (szarnekow) [inline tests/org.eclipse.xtext.xbase.ui.tests/src/org/eclipse/xtext/xbase/ui/tests/contentassist/Bug471434Test.xtend]: Really only a cosmetic thing, but the JavaDoc should appear before the annotation

File: tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2 [hunk 3]
```diff
@@ -1,11 +1,10 @@
-module org.eclipse.xtext.xbase.testlanguages.GenerateXImportSectionTestLang
+module org.eclipse.xtext.xbase.testlanguages.GenerateXbaseTestLanguages
 
 import org.eclipse.emf.mwe.utils.*
 import org.eclipse.xtext.generator.*
 import org.eclipse.xtext.ui.generator.*
 
-var grammarURI = "classpath:/org/eclipse/xtext/xbase/testlanguages/XImportSectionTestLang.xtext"
-var fileExtensions = "__ximportimporttestlang"
+var fileExtensions = ""
```
Comment (szarnekow) [inline tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2]: The file extension should be set on the languag config directly. Please stick to the extension that we had so far to reduce the number of unnecessary changes in the PR

**Comment by szarnekow** [comment]
Hey Lorenzo, thanks for the patch. Looks good to me at a first glance. Could you please regenerate Xtend, too. It would be interesting to see the impact of the change there.

File: tests/org.eclipse.xtext.xbase.testlanguages.ui/plugin.xml (hunk 1 above)
Comment (LorenzoBettini) [inline tests/org.eclipse.xtext.xbase.testlanguages.ui/plugin.xml]: that's related to a following comment

File: tests/org.eclipse.xtext.xbase.ui.tests/src/org/eclipse/xtext/xbase/ui/tests/contentassist/Bug471434Test.xtend (hunk 2 above)
Comment (LorenzoBettini) [inline tests/org.eclipse.xtext.xbase.ui.tests/src/org/eclipse/xtext/xbase/ui/tests/contentassist/Bug471434Test.xtend]: OK, I'll fix that

File: tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2 (hunk 3 above)
Comment (LorenzoBettini) [inline tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2]: I might need your help here then: what's the right way to do that (I was setting it to an empty string because I see you've already done that in another testlanguages project in the Xtext code base): If I do

```
    	language = auto-inject {
    		fileExtensions = "__ximportimporttestlang"
    		uri = "classpath:/org/eclipse/xtext/xbase/testlanguages/XImportSectionTestLang.xtext"
```

the generated standalone setup will have the method:

```
	@Override
	public List<String> getFileExtensions() {
		return Arrays.asList("__ximportimporttestlang");
	}
```

but the register method uses the wrong file extension

```
		org.eclipse.xtext.resource.IResourceServiceProvider.Registry.INSTANCE.getExtensionToFactoryMap().put("ximportsectiontestlang", serviceProvider);
```

which breaks the tests...

**Comment by LorenzoBettini** [comment]
Hi Sebastian, for the moment I only fixed the cosmetic Javadoc; concerning the languageExtensions I left a comment.
Concerning the regeneration of Xtend I haven't pushed anything yet, but that's basically the diff, which highlight the methods that are NOT generated:

```
diff --git a/plugins/org.eclipse.xtend.ide/src-gen/org/eclipse/xtend/ide/contentassist/AbstractXtendProposalProvider.java b/plugins/org.eclipse.xtend.ide/src-gen/org/eclipse/xtend/ide/contentassist/AbstractXtendProposalProvider.java
index 158cb36..a339750 100644
--- a/plugins/org.eclipse.xtend.ide/src-gen/org/eclipse/xtend/ide/contentassist/AbstractXtendProposalProvider.java
+++ b/plugins/org.eclipse.xtend.ide/src-gen/org/eclipse/xtend/ide/contentassist/AbstractXtendProposalProvider.java
@@ -162,20 +162,8 @@
 	public void completeParameter_Name(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
 	}
-	public void completeXVariableDeclaration_Writeable(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
 	public void completeXVariableDeclaration_Extension(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
-	}
-	public void completeXVariableDeclaration_Type(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeXVariableDeclaration_Name(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeXVariableDeclaration_Right(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
 	}
 	public void completeXConstructorCall_Members(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
@@ -195,32 +183,8 @@
 	public void completeJvmFormalParameter_Extension(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
 	}
-	public void completeJvmFormalParameter_ParameterType(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeJvmFormalParameter_Name(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
 	public void completeFullJvmFormalParameter_Extension(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
-	}
-	public void completeFullJvmFormalParameter_ParameterType(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeFullJvmFormalParameter_Name(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeXSwitchExpression_DeclaredParam(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeXSwitchExpression_Switch(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeXSwitchExpression_Cases(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
-	}
-	public void completeXSwitchExpression_Default(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
 	}
 	public void completeSimpleStringLiteral_Value(EObject model, Assignment assignment, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		completeRuleCall(((RuleCall)assignment.getTerminal()), context, acceptor);
@@ -328,37 +292,13 @@
 	public void complete_CreateExtensionInfo(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
 	}
-	public void complete_ValidID(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_FeatureCallID(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
 	public void complete_InnerVarID(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
 	}
 	public void complete_Parameter(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
 	}
-	public void complete_XVariableDeclaration(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_XConstructorCall(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
 	public void complete_XbaseConstructorCall(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_JvmFormalParameter(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_FullJvmFormalParameter(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_XStringLiteral(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_XSwitchExpression(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
 	}
 	public void complete_XExpressionOrSimpleConstructorCall(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
@@ -395,9 +335,6 @@
 		// subclasses may override
 	}
 	public void complete_RichStringElseIf(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
-		// subclasses may override
-	}
-	public void complete_ID(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
 		// subclasses may override
 	}
 	public void complete_HEX_DIGIT(EObject model, RuleCall ruleCall, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
```

I haven't run any tests against that yet; but the fact that the completeXVariableDeclaration_Right is _not_ generated (i.e., it does not override the XbaseProposalProvider implementation) is one of the things that fixed my DSL content assist (and also the content assist test I created for the test language)...

File: tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2 (hunk 3 above)
Comment (szarnekow) [inline tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2]: Ok - it seems to be too inconvenient to change the file extension per language without auto-inject. So the best choice seems to be the removed __ prefix.

Comment (LorenzoBettini) [inline tests/org.eclipse.xtext.xbase.testlanguages/src/org/eclipse/xtext/xbase/testlanguages/GenerateXbaseTestLanguages.mwe2]: OK, so I'm leaving it as it is in the PR.

**Comment by szarnekow** [comment]
Jenkins test

**Comment by LorenzoBettini** [comment]
Sebastian, by the way, in the Jenkins job, https://xtext-builds.itemis.de/jenkins/job/xtext-pull-request-verifier/608/consoleFull, xbase.ui.tests are not executed at all, are they?

**Comment by szarnekow** [comment]
Jenkins xtext-ide-tests-master-latest

**Comment by szarnekow** [comment]
Thanks for the hint, Lorenzo

**Comment by LorenzoBettini** [comment]
Sebastian, do you want me to further experiment with the regeneration of Xtend? Or will you take care of that?

**Comment by szarnekow** [comment]
I won't have the time right now to regenerate Xtend. If you could provide a patch, that would be helpful.

**Comment by LorenzoBettini** [comment]
OK!
